packages/core/upload/admin/src/components/SelectTree/SelectTree.js

```diff
@@ -0,0 +1,100 @@
+/* eslint-disable react/prop-types */
+
+import React, { useEffect, useState, useMemo } from 'react';
+import styled from 'styled-components';
+import PropTypes from 'prop-types';
+import { components } from 'react-select';
+
+import { Flex } from '@strapi/design-system/Flex';
+import { Icon } from '@strapi/design-system/Icon';
+import { ReactSelect as Select, pxToRem } from '@strapi/helper-plugin';
+import { Typography } from '@strapi/design-system/Typography';
+import ChevronUp from '@strapi/icons/ChevronUp';
+import ChevronDown from '@strapi/icons/ChevronDown';
+
+import flattenTree from './utils/flattenTree';
+
+const ToggleButton = styled.button`
+  align-self: flex-end;
+  margin-left: auto;
+`;
+
+const hasParent = option => !option.parent;
+
+const hasParentOrMatchesValue = (option, value) =>
+  option.value === value || option.parent === value;
+
+const SelectTree = ({ options: defaultOptions, maxDisplayDepth, ...props }) => {
+  const flatDefaultOptions = useMemo(() => flattenTree(defaultOptions), [defaultOptions]);
```

Comment:
No, they don't (except the tree will be fetched dynamically before, so there will be one render without options and one with the full tree), but I believe the `SelectTree` component will be re-rendered a couple of times in the ML (when doing validation e.g.) and since `flattenTree` can be quite expensive based on the input I thought it would be better to cache the result. Or am I missing something here?